src/test/java/org/owasp/esapi/reference/EncoderTest.java

```diff
@@ -1050,4 +1051,176 @@ public void testGetDefaultCanonicalizationCodecs() {
+            if( ch == '\b' || ch == '\f' || ch == '\r' || ch == '\n' || ch == '\t' ) {
+                continue;
+            }
+
+            final String str1 = String.format( "\\u%04x", i );
+            final String str2 = Character.toString( ch );
+            assertEquals( str1, instance.encodeForJSON(str2) );
+        }
+
+        // And the remainder of printable characters
+        for( int i=32; i<=126; ++i ) {
+            final char ch = (char)i;
+            if( ch == '/' || ch == '\\' || ch == '\"' ) {
+                continue;
+            }
+
+            final String str = Character.toString( ch );
+            assertEquals( str, instance.encodeForJSON(str) );
+        }
+    }
+
+    /**
+     * Test of decodeFromJSON method, of class org.owasp.esapi.Encoder.
+     */
+    public void testDecodeFromJSON() {
+        System.out.println("decodeFromJSON");
+        Encoder instance = ESAPI.encoder();
+
+        // Empty strings
+        assertEquals( null, instance.decodeFromJSON(null) );
+        assertEquals( "", instance.decodeFromJSON("") );
+        assertEquals( " ", instance.decodeFromJSON(" ") );
+
+        // Walk a message without escaped characters
+        String message = "Now is the time for all good men to come to the aide of their country.";
+        for( int i=1; i<message.length(); ++i ) {
+            final String substring = message.substring(0, i);
+            assertEquals( substring, instance.decodeFromJSON(substring) );
+        }
+
+        // Two-character sequence escape representations of some
+        // popular characters
+        assertEquals( "\b", instance.decodeFromJSON("\\b") );
+        assertEquals( "\f", instance.decodeFromJSON("\\f") );
+        assertEquals( "\r", instance.decodeFromJSON("\\r") );
+        assertEquals( "\n", instance.decodeFromJSON("\\n") );
+        assertEquals( "\t", instance.decodeFromJSON("\\t") );
+        assertEquals( "\"", instance.decodeFromJSON("\\\"") );
+        assertEquals( "/",  instance.decodeFromJSON("\\/" ) );
+        assertEquals( "\\", instance.decodeFromJSON("\\\\") );
+
+        // All Unicode characters may be placed within the quotation marks,
+        // except for the characters that MUST be escaped: quotation mark,
+        // reverse solidus, and the control characters (U+0000 through U+001F).
+        for( int i=0; i<=0x1f; ++i ) {
+            final String str = String.format( "\\u%04x", i );
+            final Character ch = (char)i;
+
+            assertEquals( Character.toString(ch), instance.decodeFromJSON(str) );
+        }
+
+        // And the remainder of printable characters
+        for( int i=32; i<=126; ++i ) {
+            final String str = String.format( "\\u%04x", i );
+            final Character ch = (char)i;
+
+            assertEquals( Character.toString(ch), instance.decodeFromJSON(str) );
+        }
+
+        // And a couple extra for good measure...
+        assertEquals( "\\", instance.decodeFromJSON("\\u005c") );
```

Comment:
No need to change it now, but this is an example of a test case I would want to see isolated into its own `public void testDoubleEncoded()`

Reply:
Done. All the encode and decode tests were broken out into their constituent parts.